src/content/app/tools/vep/state/vep-action-listeners/vepSubmissionStatusPolling.ts

```diff
@@ -0,0 +1,182 @@
+/**
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import config from 'config';
+
+import { updateSubmission } from 'src/content/app/tools/vep/state/vep-submissions/vepSubmissionsSlice';
+
+import type { AppDispatch } from 'src/store';
+import type { SubmissionStatus } from 'src/content/app/tools/vep/types/vepSubmission';
+
+export const POLLING_INTERVAL = 15 * 1000;
+
+type PolledSubmission = {
+  id: string;
+  status: SubmissionStatus;
+};
+
+const unsettledSubmissionsStatuses: SubmissionStatus[] = [
+  'SUBMITTED',
+  'RUNNING'
+];
+
+class VepSubmissionStatusPolling {
+  submissions: Record<string, PolledSubmission> = {};
+  queue: string[] = [];
+  isProcessing: boolean = false;
+  dispatch: AppDispatch | null = null;
+
+  enqueueSubmission = ({
+    submission,
+    dispatch
+  }: {
+    submission: PolledSubmission;
+    dispatch: AppDispatch;
+  }) => {
+    this.submissions[submission.id] = submission;
+    this.queue.unshift(submission.id);
+    this.dispatch = dispatch;
+
+    if (!this.isProcessing) {
+      this.processQueueWithDelay();
+    }
+  };
+
+  processQueueWithDelay = async () => {
+    if (!this.queue.length) {
+      this.isProcessing = false;
+      return;
+    }
+    this.isProcessing = true;
+
+    await this.pause();
+
+    const submissionId = this.queue.pop() as string;
+    const url = `${config.toolsApiBaseUrl}/vep/submissions/${submissionId}/status`;
+    const response = await fetch(url);
+    await this.handleResponse({ submissionId, response });
+
+    // this makes it a recursive function; but by avoiding the return statement, we shouldn't be leaking memory
+    this.processQueueWithDelay();
```

Comment:
So `return` happens only when the queue goes down to zero?

Reply:
No, the return happens implicitly right after line 73 executes. It doesn't wait for this internal `this.processQueueWithDelay();` to finish.